presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java

```diff
@@ -0,0 +1,543 @@
+        }
+
+        /**
+         * Drops pages up to the specified sequence id
+         */
+        private void acknowledgePages(long sequenceId)
+        {
+            checkState(!Thread.holdsLock(this), "Can not acknowledge pages while holding a lock on this");
+
+            synchronized (this) {
+                // if pages have already been acknowledged, just ignore this
+                long oldCurrentSequenceId = currentSequenceId.get();
+                if (finished.get() || sequenceId < oldCurrentSequenceId) {
+                    return;
+                }
+
+                int pagesToRemove = Ints.checkedCast(sequenceId - oldCurrentSequenceId);
+                checkState(pagesToRemove <= pages.size(), "Invalid sequence id");
+                long bytesRemoved = 0;
+                for (int i = 0; i < pagesToRemove; i++) {
+                    Page page = pages.removeFirst();
+                    bytesRemoved += page.getSizeInBytes();
+                }
+
+                verify(currentSequenceId.compareAndSet(oldCurrentSequenceId, oldCurrentSequenceId + pagesToRemove));
+
+                bufferedBytes.addAndGet(-bytesRemoved);
+                verify(bufferedBytes.get() >= 0);
+                memoryManager.updateMemoryUsage(-bytesRemoved);
+
+                // if there are no pages, and there will be no more pages, then we are finished
+                // never finish a failed buffer
+                if (pages.isEmpty() && !state.get().canAddPages() && state.get() != FAILED) {
+                    finished.set(true);
+                }
+            }
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("bufferId", bufferId)
+                    .add("sequenceId", currentSequenceId.get())
+                    .add("finished", finished.get())
+                    .toString();
+        }
+
+        @Immutable
+        private static class PendingRead
+        {
+            private final String taskInstanceId;
+            private final long sequenceId;
+            private final DataSize maxSize;
+            private final CompletableFuture<BufferResult> resultFuture = new CompletableFuture<>();
+
+            public PendingRead(String taskInstanceId, long sequenceId, DataSize maxSize)
+            {
+                this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+                this.sequenceId = sequenceId;
+                this.maxSize = maxSize;
+            }
+
+            public long getSequenceId()
+            {
+                return sequenceId;
+            }
+
+            public DataSize getMaxSize()
+            {
+                return maxSize;
+            }
+
+            public CompletableFuture<BufferResult> getResultFuture()
+            {
+                return resultFuture;
+            }
+
+            public void abort()
```

Comment:
I would inline this function. `abort` as a name is not clear to me when I read its usages. Additionally, one of the caller is wrong. It should say `true` instead of `false` for the last field. Removing `abort` will also enable removing `taskInstanceId` from this class, which doesn't really belong here.